src/main/java/com/palantir/gradle/shadowjar/ShadowJarPlugin.java

```diff
@@ -218,26 +234,27 @@ private void setupShadowJarToShadeTheCorrectDependencies(
                         .map(project.getDependencies()::create)
                         .collect(Collectors.toSet()))));
 
-        TaskProvider<ShadowJarConfigurationTask> shadowJarConfigurationTask = project.getTasks()
-                .register("relocateShadowJar", ShadowJarConfigurationTask.class, relocateTask -> {
-                    relocateTask.getShadowJar().set(shadowJarProvider.get());
+        shadowJarProvider.configure(shadowJar -> {
+            shadowJar.getConfigurations().set(shadeTransitively.map(Collections::singletonList));
 
-                    relocateTask.getPrefix().set(project.provider(() -> String.join(
-                                    ".", "shadow", project.getGroup().toString(), project.getName())
-                            .replace('-', '_')
-                            .toLowerCase(Locale.US)));
+            shadowJar.getDependencyFilter().set(shadowingCalculation.map(calc -> {
+                DefaultDependencyFilter filter = new DefaultDependencyFilter(project);
+                filter.include(calc.acceptedShadedModules()::contains);
+                return filter;
+            }));
 
-                    relocateTask.getAcceptedDependencies().set(project.provider(() -> shadowingCalculation
-                            .get()
-                            .acceptedShadedModules()));
-                });
+            String prefix = String.join(".", "shadow", project.getGroup().toString(), project.getName())
+                    .replace('-', '_')
+                    .toLowerCase(Locale.US);
 
-        shadowJarProvider.configure(shadowJar -> {
-            shadowJar.dependsOn(shadowJarConfigurationTask);
-            shadowJar.getConfigurations().set(shadeTransitively.map(Collections::singletonList));
+            Provider<FileCollection> includedDepsProvider = project.provider(shadowJar::getIncludedDependencies);
+
+            Provider<Set<String>> pathsInJars = includedDepsProvider.map(ShadowJarPlugin::scanJarsForPaths);
+
+            Provider<Set<String>> relocatablePaths = pathsInJars.map(ShadowJarPlugin::computeRelocatablePaths);
+
+            shadowJar.getRelocators().add(new JarFilesRelocator(relocatablePaths, prefix + "."));
```

Comment:
I'm wondering whether we can side step a bunch of the weird `JarFilesRelocator` shenanigans and instead just make use of `SimpleRelocator`. The implementation seems to do most of what we want it to do! So what if we could basically do:

```java
shadowJar.getRelocators()
  .add(relocatablePaths.map(paths -> new SimpleRelocator("", prefix + ".", paths, ImmutableSet.of()));
```

The current impl (at least in 9.2.2) handles the class suffix, and some more bits i.e. just comparing the `canRelocatePath` implementations. We already don't do anything in `relocateClass` and `relocatePath`, and at this point might just be able to delete the file. 

I think this is worth testing however with a bunch of internal repos. But it will simplify things a lot I think.